Shoko.Server/ImageDownload/ImageDownloadRequest.cs

```diff
@@ -1,17 +1,126 @@
+            AnimeStaff staff => staff.GetImagePath(),
+            MovieDB_Fanart image => image.GetFullImagePath(),
+            MovieDB_Poster image => image.GetFullImagePath(),
+            SVR_AniDB_Anime anime => anime.PosterPath,
+            TvDB_Episode episode => episode.GetFullImagePath(),
+            TvDB_ImageFanart image => image.GetFullImagePath(),
+            TvDB_ImagePoster image => image.GetFullImagePath(),
+            TvDB_ImageWideBanner image => image.GetFullImagePath(),
+            _ => string.Empty,
+        };
+
+    private string? _downloadUrl { get; set; } = null;
+
+    public string DownloadUrl
+        => _downloadUrl != null ? _downloadUrl : _downloadUrl = ImageData switch
+        {
+            AniDB_Character character => string.Format(ImageServerUrl, character.PicName),
+            AniDB_Seiyuu creator => string.Format(ImageServerUrl, creator.PicName),
+            AnimeStaff staff => string.Format(ImageServerUrl, staff.ImagePath),
+            MovieDB_Fanart movieFanart => string.Format(Constants.URLS.MovieDB_Images, movieFanart.URL),
+            MovieDB_Poster moviePoster => string.Format(Constants.URLS.MovieDB_Images, moviePoster.URL),
+            SVR_AniDB_Anime anime => string.Format(ImageServerUrl, anime.Picname),
+            TvDB_Episode ep => string.Format(Constants.URLS.TvDB_Episode_Images, ep.Filename),
+            TvDB_ImageFanart fanart => string.Format(Constants.URLS.TvDB_Images, fanart.BannerPath),
+            TvDB_ImagePoster poster => string.Format(Constants.URLS.TvDB_Images, poster.BannerPath),
+            TvDB_ImageWideBanner wideBanner => string.Format(Constants.URLS.TvDB_Images, wideBanner.BannerPath),
+            _ => string.Empty
+        };
+
+    public ImageDownloadRequest(object data, string? imageServerUrl = null)
     {
-        ImageType = imageType;
         ImageData = data;
-        ForceDownload = forceDownload;
+        ImageServerUrl = imageServerUrl ?? "";
+    }
+
+    public string? DownloadNow(int maxRetries = 5)
+        => RecursivelyRetryDownload(0, maxRetries);
+
+    private string? RecursivelyRetryDownload(int count, int maxRetries)
+    {
+        // Abort if the download url or final destination is not available.
+        if (string.IsNullOrEmpty(DownloadUrl) || string.IsNullOrEmpty(FilePath))
+            return null;
+
+        var tempPath = Path.Combine(ImageUtils.GetImagesTempFolder(), Path.GetFileName(FilePath));
+        try
+        {
+            // Ignore all certificate failures.
+            ServicePointManager.Expect100Continue = true;
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+            ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
+
+            // Download the image.
+            using (var client = new HttpClient())
+            {
+                // Download the image data.
+                client.DefaultRequestHeaders.Add("user-agent", "JMM");
```

Comment:
Should we update or keep the "JMM" user agent?

Reply:
Keep. AniDB whitelists the user agent for some things, and I don't want to fuck with it